### DISCUSSION TITLE
Validation Attributes

### DISCUSSION THREAD

**Discussion by markhuot** [laravel/framework#53784] [Ideas]
I'm loving the new `#[CurrentUser]` attribute in the HTTP layer. I love having strongly typed data coming in to my controller and it's making me wonder if it'd be possible to do the same with validated form data. Something along the lines of,

```php

class CreateTodoController
{
  public function __invoke(
    #[CurrentUser] $user,
    #[Validated('string|required')] string $title,
    #[Validated('string')] ?string $description=null,
    #[Validated('date')] ?DateTime $dueDate=null,
  ) {
    $todo = new Todo;
    $todo->title = $title;
    $todo->description = $description;
    $todo->dueDate = $dueDate;
    $todo->save();

    return redirect()->route('dashboard.index');
  }
}
```

Today I do this same sort of thing with https://github.com/spatie/laravel-data and absolutely _love_ it. But, if I could avoid all the extra value objects that are really only being used as strongly typed `FormRequest`s then I think I'd be a lot happier.

I'm looking for hooks in to the container resolution where I could do something like this and I'm not seeing anything. Anyone know if it's possible to resolve a group of parameters before Laravel does its thing?

----

Edited to add that we could use some of that `laravel-data` type inferring to turn a `string` type hint in to a rule. That way you could drop a lot of that noise,

```php
public function __invoke(
  #[CurrentUser] $user,
  #[Validated] string $title,
  #[Validated] ?string $description=null,
  #[Validated] ?DateTime $dueDate=null,
) {
```

**Comment by shaedrich** [comment]
I'd love to have that! 😍

**Comment by macropay-solutions** [comment]
How would be handled the validation exceptions in this case?
